### ISSUE TITLE
Impossible to make the formatting toolbar appear after Escape has been pressed

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#5501]
To reproduce, use only the keyboard:

Navigate to a block and select some text: the formatting toolbar appears

<img width="697" alt="screen shot 2018-03-08 at 13 07 10" src="https://user-images.githubusercontent.com/1682452/37150533-6292b224-22d2-11e8-9cc9-e975ab9e9028.png">

Shift + Tab to go to the toolbar and apply some formatting:

<img width="701" alt="screen shot 2018-03-08 at 13 07 15" src="https://user-images.githubusercontent.com/1682452/37150572-8314b2a4-22d2-11e8-9e14-fefc3fa68fd2.png">

Press Escape: the toolbar correctly disappears and focus is on the block again:

<img width="656" alt="screen shot 2018-03-08 at 13 07 26" src="https://user-images.githubusercontent.com/1682452/37150604-9c56f7fe-22d2-11e8-972d-58a528dc64d2.png">

Use the arrow keys to move the cursor to some other text and Shift + arrow to select some text (same if you select the previous text):

<img width="669" alt="screen shot 2018-03-08 at 13 07 33" src="https://user-images.githubusercontent.com/1682452/37150662-ce8d664a-22d2-11e8-97db-c58d5f5f3759.png">

The formatting toolbar doesn't appear.  The only way I've found to "fix" this is to tab away from teh block and then tab back into it.

Expected: when there's a text selection, the formatting toolbar should always appear.

Suspect: pressing Escape propagates to the block and the block is no more "selected" (notice in the screenshot above the block movers and more button are not shown).

**Comment by afercia** [comment]
Note: this happens even without going to the toolbar to apply some formatting. Just:
- select some text
- press Escape
- select again some text: the toolbar doesn't appear

**Comment by aduth** [comment]
This should be somewhat improved after #5513.

Specifically the notes on pressing escape in a block:

>Pressing escape in a block should do nothing
> - Arguable whether this is the desired behavior, but it is consistent with native inputs. Master is in a worse state because the block becomes deselected but the text caret is not removed, causing issues because the block never again becomes selected (because `isSelected` is synced to its `focus` event, but focus has never left).

Pressing escape in the toolbar has slightly different behavior, because we're annoyingly handling escape keys in many places in many different ways.

The NavigableToolbar is shifting focus to the selected block:

https://github.com/WordPress/gutenberg/blob/4cb41ab3113070cca81f17c11198333f12f9ba8d/editor/components/navigable-toolbar/index.js#L35-L47

Even though `EditorGlobalKeyboardShortcuts` is simultaneously clearing the selected block:

https://github.com/WordPress/gutenberg/blob/4cb41ab3113070cca81f17c11198333f12f9ba8d/editor/components/editor-global-keyboard-shortcuts/index.js#L77

So the paragraph block border shows, but the block is not technically selected. Pressing tab once more returns to the selection which existed prior to the escape press (and the block is once again selected).

Maybe intended / desirable, not sure. It all seems rather convoluted and fragile though in its current form.

**Comment by afercia** [comment]
I'd agree it's a bit fragile right now and maybe needs some rethinking.

Pressing Escape in the toolbar should just return focus to the edited block still in a "selected" state, as I'm still editing the block.

Re: #5513 
> Pressing escape in a block should do nothing

Worth considering the pending experiment about "navigation mode / edit mode", see https://github.com/WordPress/gutenberg/pull/3195 and https://github.com/WordPress/gutenberg/issues/2031. If implemented, then pressing Escape in a block should switch back to "navigation mode."
